docs/specs/alcproxies.md

```diff
@@ -0,0 +1,251 @@
+Making the `ClientObject` a `DispatchProxy` object would work for making calls over ALCs, though not for out-of-process calls. It's not the perfect solution, as it would require two `Reflection.Invoke` calls (one from client to server, and one from server to target), but it would allow us to enforce our isolation policy from the custom `ServerObject` being able to mess with our return calls to make sure we were enforcing our constraints (see "ALC Unloading" below).
+
+#### Binary Serialization
+Using `BinaryFormatter`, types and their information can be serialized and sent across the ALC barrier, and deserialized using a class for decoding, which can then call the specific `TargetObject` methods.
+
+Benefits to using `BinaryFormatter`:
+* It is usable with any objects, as long as they are given the [Serializable] attribute.
+* There's no need to re-implement serialization, which would probably be out of scope for this project.
+
+Concerns with using `BinaryFormatter`:
+* Security concerns, as untrusted sources can use serialization to gain unwanted access to a program (there are a few examples of this [here](https://media.blackhat.com/bh-us-12/Briefings/Forshaw/BH_US_12_Forshaw_Are_You_My_Type_WP.pdf )).
+* It requires any objects that are being piped through the API to have the [Serializable] attribute, which needs to be manually done.
+* It's uncertain how we would use this for inter-process communication.
+* Potential casting issues which could prevent non-primitives from being passed through correctly.
+
+#### XML Serialization using `DataContractSerializer`
+Instead of moving a specific type through a `BinaryFormatter`, using `DataContractSerializer` would allow us to move any public or private pieces of an object, and then recreate it on the other side.
+
+Benefits of `DataContractSerializer`:
+* It may be easier compared to `BinaryFormatter` to cast this effectively across the type-isolation barrier.
+* Adding attributes is similar to `BinaryFormatter`, so any type should be able to make it across the barrier.
+
+Concerns with `DataContractSerializer`:
+* Many similar concerns to `BinaryFormatter`, such as potential security issues and the manual use of [DataContractAttribute] and [DataMemberAttribute] needing to be used.
+
+#### gRPC
+Using the [gRPC](https://grpc.io/ ) framework, we'll have a much easier time simulating inter-process communication, as it's a major part of the framework.
+
+Benefits of gRPC:
+* Easy inter-process communication.
+* Simple easy to implement framework, with [some support for .NET](https://github.com/grpc/grpc-dotnet ).
+
+Concerns with gRPC:
+* Encoding objects in a way that we can manually decode and pass them into the `TargetObject` needs to be investigated further, to make sure it works.
+* At the very least, there will probably be more overhead with how gRPC does in-process communication calls. At worst, it won't work since most of it is designed for use with ASP.NET instead of just .NET Core.
```

Comment:
The gRPC itself will work anywhere... the .NET implementation is bound to HTTP transfer (at least that's what I've seen so far) which would be a problem for in-proc communication.

Reply:
I know there was some interest on their repos for in-proc use, but it seems like grpc-dotnet doesn't have anything for that right now (closest thing I could find was [this](https://github.com/grpc/grpc-dotnet/issues/239 ), so I can probably remove all the in-proc references to using gRPC.

Reply:
Agreed.